backend/FwLite/FwLiteShared/Projects/CombinedProjectsService.cs

```diff
@@ -143,12 +155,38 @@ private ProjectRole FromRole(UserProjectRole role) =>
             _ => ProjectRole.Unknown
         };
 
-    public async Task DownloadProject(string code, LexboxServer server)
+    [JSInvokable]
+    public async Task<DownloadProjectByCodeResult> DownloadProjectByCode(string code, LexboxServer server, string? userRole = null)
     {
         var serverProjects = await ServerProjects(server, false);
-        var project = serverProjects.FirstOrDefault(p => p.Code == code)
-            ?? throw new InvalidOperationException($"Project {code} not found on server {server.Authority}");
+        var project = serverProjects.Projects.FirstOrDefault(p => p.Code == code);
+        if (project is null)
+        {
+            if (serverProjects.CanDownloadProjectsWithoutMembership)
+            {
+                var projectId = await lexboxProjectService.GetLexboxProjectId(server, code);
+                if (projectId is null) return DownloadProjectByCodeResult.ProjectNotFound;
+                var isCrdtProject = await lexboxProjectService.IsCrdtProject(server, code);
+                if (!isCrdtProject) return DownloadProjectByCodeResult.NotCrdtProject;
+                var role = ProjectRole.Editor;
+                if (userRole is not null) Enum.TryParse(userRole, out role);
+                project = new ProjectModel(
+                    Name: code,
+                    Code: code,
+                    Crdt: isCrdtProject,
+                    Fwdata: false,
+                    Lexbox: true,
+                    Role: role,
+                    Server: server,
+                    Id: projectId
+                );
+                await DownloadProject(project);
+                return DownloadProjectByCodeResult.Success;
+            }
+            return DownloadProjectByCodeResult.ProjectNotFound;
+        }
         await DownloadProject(project);
+        return DownloadProjectByCodeResult.Success;
```

Comment:
shouldn't there be a check here that the project already exists? or does that happen on the frontend?

Reply:
There's a check on the frontend that won't let you submit the dialog if the project already exists in localProjects, but a backend check is probably a good idea also in case of frontend bugs.